odxtools/request.py

```diff
@@ -1,39 +1,122 @@
+from .decodestate import DecodeState
+from .element import IdentifiableElement
 from .encodestate import EncodeState
-from .odxlink import OdxDocFragment
-from .odxtypes import ParameterValue
+from .exceptions import odxraise
+from .nameditemlist import NamedItemList
+from .odxlink import OdxDocFragment, OdxLinkDatabase, OdxLinkId
+from .odxtypes import ParameterValue, ParameterValueDict
+from .parameters.createanyparameter import create_any_parameter_from_et
+from .parameters.parameter import Parameter
 from .snrefcontext import SnRefContext
+from .specialdatagroup import SpecialDataGroup
 from .utils import dataclass_fields_asdict
 
 
-# TODO: The spec does not say that requests are basic structures. For
-# now, we derive from it anyway because it simplifies the en- and
-# decoding machinery...
 @dataclass
-class Request(BasicStructure):
+class Request(IdentifiableElement):
+    """Represents all information related to an UDS request
+
+    This class implements the `CompositeCodec` and `ToplevelCodec`
+    interfaces.
+
+    """
+    admin_data: Optional[AdminData]
+    parameters: NamedItemList[Parameter]
+    sdgs: List[SpecialDataGroup]
+
+    @property
+    def required_parameters(self) -> List[Parameter]:
+        return composite_codec_get_required_parameters(self)
+
+    @property
+    def free_parameters(self) -> List[Parameter]:
+        return composite_codec_get_free_parameters(self)
+
+    def __post_init__(self) -> None:
+        assert isinstance(self, CompositeCodec)
+        assert isinstance(self, ToplevelCodec)
 
     @staticmethod
     def from_et(et_element: ElementTree.Element, doc_frags: List[OdxDocFragment]) -> "Request":
-        """Reads a response."""
-        kwargs = dataclass_fields_asdict(BasicStructure.from_et(et_element, doc_frags))
+        kwargs = dataclass_fields_asdict(IdentifiableElement.from_et(et_element, doc_frags))
+
+        admin_data = AdminData.from_et(et_element.find("ADMIN-DATA"), doc_frags)
+        parameters = NamedItemList([
+            create_any_parameter_from_et(et_parameter, doc_frags)
+            for et_parameter in et_element.iterfind("PARAMS/PARAM")
+        ])
+        sdgs = [
+            SpecialDataGroup.from_et(sdge, doc_frags) for sdge in et_element.iterfind("SDGS/SDG")
+        ]
+
+        return Request(admin_data=admin_data, parameters=parameters, sdgs=sdgs, **kwargs)
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        result = {self.odx_id: self}
+
+        for param in self.parameters:
+            result.update(param._build_odxlinks())
+
+        return result
+
+    def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+        for param in self.parameters:
+            param._resolve_odxlinks(odxlinks)
 
-        return Request(**kwargs)
+    def _resolve_snrefs(self, context: SnRefContext) -> None:
+        context.request = self
+        context.parameters = self.parameters
+
+        try:
```

Comment:
Why try? it's not expected to raise

Reply:
this is a defensive measure to make it not screw up more than necessary even if we are in non-strict mode and an exception is raised by a parameter's `.resolve_snref()`

Reply:
code comment to explain would be nice then

Reply:
come think about it, you're right: this code can be removed. fixed: https://github.com/mercedes-benz/odxtools/pull/364/commits/e5d21e67e16a72abf25a16b3aa2ed07b868e15e6